File: samples/sample-app-java/src/main/java/com/weedow/searchy/sample/java/entity/Vehicle.java
package com.weedow.searchy.sample.java.entity;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import javax.persistence.*;
import java.util.Map;

@Entity
public class Vehicle extends JpaPersistable<Long> {

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private VehicleType vehicleType;

    @Column(nullable = false)
    private String brand;

    @Column(nullable = false)
    private String model;

    @ManyToOne(optional = false)
    @JsonIgnoreProperties({"vehicles"})
    private Person person;

    @OneToMany(cascade = {CascadeType.ALL})
    @JoinTable(name = "feature_mapping",
            joinColumns = {@JoinColumn(name = "vehicle_id", referencedColumnName = "id")},
            inverseJoinColumns = {@JoinColumn(name = "feature_id", referencedColumnName = "id")})
    @MapKey(name = "name") // Feature name
    private Map<String, Feature> features;

    public VehicleType getVehicleType() {
        return vehicleType;
    }

    public Vehicle setVehicleType(VehicleType vehicleType) {
        this.vehicleType = vehicleType;
        return this;
    }

    public String getBrand() {
        return brand;
    }

    public Vehicle setBrand(String brand) {
        this.brand = brand;
        return this;
    }

    public String getModel() {
        return model;
    }

    public Vehicle setModel(String model) {
        this.model = model;
        return this;
    }

    public Person getPerson() {
        return person;
    }

    public Vehicle setPerson(Person person) {
        this.person = person;
        return this;
    }

    public Map<String, Feature> getFeatures() {
        return features;
    }

    public Vehicle setFeatures(Map<String, Feature> features) {
        this.features = features;
        return this;
    }
}

File: common/src/main/kotlin/com/weedow/spring/data/search/common/model/Vehicle.kt
package com.weedow.spring.data.search.common.model

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import javax.persistence.*

@Entity
class Vehicle(
        @Column(nullable = false)
        @Enumerated(EnumType.STRING)
        val vehicleType: VehicleType,

        @Column(nullable = false)
        val brand: String,

        @Column(nullable = false)
        val model: String,

        @ManyToOne(optional = false)
        @JsonIgnoreProperties("vehicles")
        val person: Person,

        @OneToMany(cascade = [CascadeType.ALL])
        @JoinTable(name = "feature_mapping",
                joinColumns = [JoinColumn(name = "vehicle_id", referencedColumnName = "id")],
                inverseJoinColumns = [JoinColumn(name = "feature_id", referencedColumnName = "id")])
        @MapKey(name = "name") // Feature name
        val features: Map<String, Feature>? = null

) : JpaPersistable<Long>()

enum class VehicleType {
    CAR, MOTORBIKE, SCOOTER, VAN, TRUCK
}
